File: Command/Command.Py/src/remote_control.py
from . import command
from .command import Action


class RemoteControl:

    def __init__(self):
        self.undo_command: Action = command.null_command()
        self._on_commands: list[Action] = [command.null_command() for _ in range(7)]
        self._off_commands: list[Action] = [command.null_command() for _ in range(7)]

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}"

    def __str__(self) -> str:
        string: str = "\n------ Remote Control -------\n"
        for i in range(7):
            string += ( f"[slot {i}] {self._on_commands[i].__name__} "
                        f"{self._off_commands[i].__name__} \n" )
        string += f"[undo] {self.undo_command.__name__} \n"
        return string

    @property
    def on_commands(self) -> list[Action]:
        return self._on_commands

    @on_commands.setter
    def on_commands(self) -> None:
        raise AttributeError("'on_commands' is read only")

    @property
    def off_commands(self) -> list[Action]:
        return self._off_commands

    @off_commands.setter
    def off_commands(self) -> None:
        raise AttributeError("'off_commands' is read only")

    def set_command(self, slot: int, on_command: Action, off_command: Action) -> None:
        self._on_commands[slot] = on_command
        self._off_commands[slot] = off_command

    def press_on_button(self, slot: int) -> None:
        self._on_commands[slot]()
        self.undo_command = self._on_commands[slot]

    def press_off_button(self, slot: int) -> None:
        self._off_commands[slot]()
        self.undo_command = self._off_commands[slot]

    def press_undo_button(self) -> None:
        self.undo_command(undo=True)

File: Command/DotNet/Command/RemoteControl.cs
using CommandPattern;

namespace CommandPattern.Controls;


public class SimpleRemoteControl
{
    public Command Slot;

    public SimpleRemoteControl(){}

    public void SetCommand(Command command)
    {
        this.Slot = command;
    }

    public void ButtonWasPressed()
    {
        this.Slot.Execute();
    }
}

// OBJECT ORIENTED WAY
public class RemoteControl
{
    public Command[] OnCommands { get; }
    public Command[] OffCommands { get; }
    public Command UndoCommand { get; set; } = new NullCommand();

    public RemoteControl()
    {
        this.OnCommands = new Command[7];
        this.OffCommands = new Command[7];
        for (int i = 0; i < 7; i++)
        {
            this.OnCommands[i] = new NullCommand();
            this.OffCommands[i] = new NullCommand();
        }
    }

    public void SetCommand(int slot, Command onCommand, Command offCommand)
    {
        this.OnCommands[slot] = onCommand;
        this.OffCommands[slot] = offCommand;
    }

    public void OnButtonWasPressed(int slot)
    {
        this.OnCommands[slot].Execute();
        this.UndoCommand = this.OnCommands[slot];
    }

    public void OffButtonWasPressed(int slot)
    {
        this.OffCommands[slot].Execute();
        this.UndoCommand = this.OffCommands[slot];
    }

    public void UndoButtonWasPressed()
    {
        this.UndoCommand.Undo();
    }

    public override string ToString()
    {
        string buffer = "\n------ Remote Control -------\n";
        for (int i = 0; i < this.OnCommands.Length; i++)
            buffer += $"[slot {i}] {this.OnCommands[i].GetType().Name} "
                    + $"{this.OffCommands[i].GetType().Name} \n";
        buffer += $"[undo] {this.UndoCommand.GetType().Name} \n";
        return buffer;
    }
}


// // FUNCTIONAL(ISH) APPROACH
// public class RemoteControl
// {
//     public Action[] OnCommands { get; }
//     public Action[] OffCommands { get; }

//     public RemoteControl()
//     {
//         this.OnCommands = new Action[7];
//         this.OffCommands = new Action[7];
//         for (int i = 0; i < 7; i++)
//         {
//             this.OnCommands[i] = () => {};
//             this.OffCommands[i] = () => {};
//         }
//     }

//     public void SetCommand(int slot, Action onCommand, Action offCommand)
//     {
//         this.OnCommands[slot] = onCommand;
//         this.OffCommands[slot] = offCommand;
//     }

//     public void OnButtonWasPressed(int slot)
//     {
//         this.OnCommands[slot]();
//     }

//     public void OffButtonWasPressed(int slot)
//     {
//         this.OffCommands[slot]();
//     }

//     public override string ToString()
//     {
//         string buffer = "\n------ Remote Control -------\n";
//         for (int i = 0; i < this.OnCommands.Length; i++)
//             buffer += $"[slot {i}] {this.OnCommands[i].GetType().Name} "
//                     + $"{this.OffCommands[i].GetType().Name} \n";
//         return buffer;
//     }
// }